### PR TITLE
feat: Support implicit and named relations.

### PR DESCRIPTION
**Pull request by Isakdl**
# Changes

- Introduces the `name` keyword inside a relation. If used it is expected that there is a named relation on both sides. The name can be an arbitrary string, but it has to be unique per file / relation.
- Introduces implicit 1:n relations. If only the list side is defined a relation is created behind the scene by injecting a hidden foreign key into the database.
- Strict validation for the above rules
- Validate that onDelete and onUpdate cannot be used on the side without the foreign key.

## Examples

In the 1:1 case I have committed t

**1:1, both sides defined**
```yaml
# user.yaml
class: User
table: user
fields:
  addressId: int
  address: Address?, relation(name=user_address, field=addressId)
indexes:
  user_address_unique_idx:
    fields: addressId
    unique: true

# address.yaml
class: Address
table: address
fields:
  user: User?, relation(name=user_address)
```

In the above situation everything is explicitly defined, no id field is created on the address side, only the user side has information about the relation in the database, `addressId` holds the foreign key. Here it is __required__ to specify the field as we cannot automatically determine which side should the foreign key.

**1:1, object on one side, id on the other**
```yaml
# user.yaml
class: User
table: user
fields:
  addressId: int, relation(name=user_address, parent=address)
indexes:
  user_address_unique_idx:
    fields: addressId
    unique: true

# address.yaml
class: Address
table: address
fields:
  user: User?, relation(name=user_address)
```

This relation is the same as the first example, but we do not have a reference to the address object in the user, we only have access to the id. On the address side we have access to the entire user object.

**1:1, two independent relations**
```yaml
# user.yaml
class: User
table: user
fields:
  address: Address?, relation

# address.yaml
class: Address
table: address
fields:
  user: User?, relation
```

Before this PR the above example worked as the first example does now. With this PR this changes to instead create two independent relations. On the user side an `addressId` field is injected holding the primary key, and on the address side a `userId` field is injected holding the primary key.

**Explicit 1:n object on one side, list on the other**
```yaml
# employee.yaml
class: Employee
table: employee
fields:
  company: Company?, relation(name=company_employees)


# company.yaml
class: Company
table: company
fields:
  employees: List<Employee>?, relation(name=company_employees)
```
In this situation a `companyId` field is injected in the employee and holds the foreign key in the db.

**Explicit 1:n id on one side, list on the other**
```yaml
# employee.yaml
class: Employee
table: employee
fields:
  companyId: int, relation(name=company_employees)


# company.yaml
class: Company
table: company
fields:
  employees: List<Employee>?, relation(name=company_employees)
```
In this situation the `companyId` holds the foreign key.

**1:n only one side defined**
```yaml
# employee.yaml
class: Employee
table: employee
fields:
  companyId: int, relation(parent=company)

# company.yaml
class: Company
table: company
fields:
  name: String
```

The above example is a valid 1:n relation but the company holds no list of the employees. The only real difference between a 1:1 and a 1:n is that in a 1:1 the foreign key needs to be unique. 

**Implicit 1:n**
```yaml
# employee.yaml
class: Employee
table: employee
fields:
  name: String


# company.yaml
class: Company
table: company
fields:
  employees: List<Employee>?, relation
```

In the above example we create a hidden column in the employee table that holds the foreign key. 

**1:1 self relation**
```yaml
class: Post
  table: post
  fields:
    content: String
    previous: Post?, relation(name=next_previous_post)
    nextId: int?
    next: Post?, relation(name=next_previous_post, field=nextId)
indexes:
  post_next_id_idx:
    fields: nextId
    unique: true
```

**1:n self relation**
```yaml
class: User
  table: user
  fields:
    name: String
    parent: User?, relation(name=parent_child, optional)
    children: List<User>?, relation(name=parent_child)
```

Discussion point on this one, here the id field has to logically be optional or else you will not be able to create the first entry. In the case of an implicit foreign key field do we want to instead automatically set it to optional? If the user instead would define the field manually would we want to lint that it has to be optional?

## Related issues
Closes: https://github.com/serverpod/serverpod/issues/1231
Closes: https://github.com/serverpod/serverpod/issues/1239

## Limitations

This PR implements the protocol specification side of the logic, I.E. defining the database and model structures. Support for insert, update, delete and read operations are to be dealt with in subsequent PR(s).

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

## Breaking changes

Breaking change to implementation on main!
Previously we linked up relations on both sides when they where not named. This situation will now result in two independent relations instead. To get the same behavior as previously you have to specify a name on the relation on both sides.

E.g. `relation(name=a_name_for_the_relation)`

### CODE REVIEW

File: tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_many_test.dart [hunk 1]
```diff
@@ -205,18 +322,220 @@ fields:
...
+
+      expect(
+        errors.last.message,
+        'The "name" property is mutually exclusive with the "parent" property.',
+      );
+    }, skip: errors.isEmpty);
+  });
+
+  group(
+      'Given a class with a one to many relation where the relationship where is named with a none string value',
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_many_test.dart:393]: NIT:
```suggestion
      'Given a class with a one to many relation where the relationship is named with a none string value',
```

File: tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_one_test.dart [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+    group('then the company field has a relation', () {
+      var relation = userDefinition.findField('address')?.relation;
+      test('of ObjectRelation type', () {
+        expect(
+          relation.runtimeType,
+          ObjectRelationDefinition,
+        );
+      });
+
+      test('with out a name for the relation', () {
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_one_test.dart:172]: NIT:
```suggestion
      test('without a name for the relation', () {
```

File: tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_one_test.dart [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+      definition1,
+      entities,
+    );
+    SerializableEntityAnalyzer.validateYamlDefinition(
+      protocol2.yaml,
+      protocol2.yamlSourceUri.path,
+      collector,
+      definition2,
+      entities,
+    );
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_one_test.dart:379]: Suggestion: Move this into a helper method please 🙏 

After reading a bunch of these tests this boilerplate just makes the test harder to understand. It separates the tests cases from the definitions that are being validated. I don't think it helps with the understanding of the test and doesn't really reflect the abstraction level we are testing on.

```suggestion
    var definition1 = SerializableEntityAnalyzer.extractEntityDefinition(
      protocol1,
    );

    var definition2 = SerializableEntityAnalyzer.extractEntityDefinition(
      protocol2,
    );

    var entities = [definition1!, definition2!];

    SerializableEntityAnalyzer.resolveEntityDependencies(entities);

    SerializableEntityAnalyzer.validateYamlDefinition(
      protocol1.yaml,
      protocol1.yamlSourceUri.path,
      collector,
      definition1,
      entities,
    );
    SerializableEntityAnalyzer.validateYamlDefinition(
      protocol2.yaml,
      protocol2.yamlSourceUri.path,
      collector,
      definition2,
      entities,
    );
```

**Review by SandPod — COMMENTED** [review]
Nice work! 🚀 

Nice to get this in, have a couple of questions and suggestions.

**Review by vlidholt — APPROVED** [review]
LGTM

File: tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_one_test.dart (hunk 3 above)
Comment (Isakdl) [reply tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_one_to_one_test.dart:379]: Will do in another PR

**Pull request merged**